linker/shared/src/main/scala/org/scalajs/linker/backend/webassembly/TextWriter.scala

```diff
@@ -0,0 +1,611 @@
+          b.sameLineList(
+            "func",
+            { appendName(id) }
+          )
+        case ExportDesc.Global(id) =>
+          b.sameLineList(
+            "global",
+            { appendName(id) }
+          )
+      }
+    }
+  )
+
+  private def writeStart(startFunction: FunctionID)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "start", {
+        appendName(startFunction)
+      }
+    )
+  }
+
+  private def writeElement(element: Element)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "elem", {
+        element.mode match {
+          case Element.Mode.Declarative => b.appendElement("declare")
+        }
+        writeType(element.tpe)
+        element.init.foreach { item =>
+          b.newLineList(
+            "item",
+            item.instr.foreach(writeInstr(_))
+          )
+        }
+      }
+    )
+  }
+
+  private def writeData(data: Data)(implicit b: WatBuilder): Unit = {
+    b.newLineList(
+      "data", {
+        appendName(data.id)
+        data.mode match {
+          case Data.Mode.Passive => ()
+        }
+        b.appendElement("\"" + data.bytes.map("\\%02x".format(_)).mkString + "\"")
+      }
+    )
+  }
+
+  private def writeTypeUse(typeID: TypeID)(implicit b: WatBuilder): Unit = {
+    b.sameLineList("type", appendName(typeID))
+  }
+
+  private def writeType(tpe: StorageType)(implicit b: WatBuilder): Unit = {
+    tpe match {
+      case tpe: SimpleType => b.appendElement(tpe.textName)
+      case tpe: PackedType => b.appendElement(tpe.textName)
+
+      case RefType(true, heapType: HeapType.AbsHeapType) =>
+        b.appendElement(heapType.nullableRefTextName)
+
+      case RefType(nullable, heapType) =>
+        b.sameLineList(
+          "ref", {
+            if (nullable)
+              b.appendElement("null")
+            writeHeapType(heapType)
+          }
+        )
+    }
+  }
+
+  private def writeHeapType(heapType: HeapType)(implicit b: WatBuilder): Unit = {
+    heapType match {
+      case HeapType.Type(typeID)          => appendName(typeID)
+      case heapType: HeapType.AbsHeapType => b.appendElement(heapType.textName)
+    }
+  }
+
+  private def floatString(v: Double): String = {
+    if (v.isNaN()) "nan"
+    else if (v == Double.PositiveInfinity) "inf"
+    else if (v == Double.NegativeInfinity) "-inf"
+    else if (v.equals(-0.0)) "-0.0"
+    else v.toString()
+  }
+
+  private def writeBlockType(blockType: BlockType)(implicit b: WatBuilder): Unit = {
+    blockType match {
+      case BlockType.FunctionType(typeID) =>
+        writeTypeUse(typeID)
+      case BlockType.ValueType(optType) =>
+        for (tpe <- optType)
+          b.sameLineList("result", writeType(tpe))
+    }
+  }
+
+  private def writeLabelIdx(labelIdx: LabelID)(implicit b: WatBuilder): Unit =
+    appendName(labelIdx)
+
+  private def writeInstr(instr: Instr)(implicit b: WatBuilder): Unit = {
+    instr match {
+      case PositionMark(_) =>
+        // ignore
+        ()
```

Comment:
I'm not sure that in general in the SJS codebase we write out units explicitly like this.

But you are the code style arbiter :)